### PR TITLE
[QMS-503] fix brouter local setup

### PR DESCRIPTION
**Pull request by ntruchsess**
### What is the linked issue for this pull request:

QMS-#503

### What you have done:
adjusted the local brouter-setup wizard to work with the new zip-file format of brouter-version >=1.6.3

### Steps to perform a simple smoke test:
Open BRouter->Setup->local installation->update existing BRouter installation->brouter-1.6.3.zip->Download and install->(wait for installation to finish->next ...

### Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):

- [X] yes

### Is every user facing string in a tr() macro?

- [X] yes

### Did you add the ticket number and title into the changelog? Keep the numeric order in each release block.

- [X] yes, I didn't forget to change changelog.txt

### CODE REVIEW

**Review by kiozen — CHANGES_REQUESTED** [review]
Now I get:
```
Error:
cd /home/oeichler/Dokumente/QMapShack/BRouter
/usr/bin/java -Xmx128M -Xms128M -Xmn8M -DmaxRunningTime=300 -cp brouter-1.6.3-all.jar btools.server.RouteServer segments4 profiles2 customprofiles 17777 1
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" java.lang.UnsupportedClassVersionError: btools/server/RouteServer has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:473)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:601)
```
Wrong Java version. Yay! Why does Java suck so much?

Anyways can you please take care of all the needless copies and review the code code for more of them? I am sure I didn't catch all ;)

And I try to find out how to fix that Java bullshit.

**Comment by kiozen** [comment]
Ok, installing a more recent OpenJDK version and pointing to the java binary worked. I fear this is annoying many other users, too. Either dropping QMapShack right from the spot or filing bug reports. Would it be possible to check this during installation and fail with a short non java geeky error message if the java version does not match?

**Comment by wthaem** [comment]
The java version issue has been mentioned in #503. There are different ways to find the version of the java installation and the java version used to compile the jar files. If they can be used with all operating systems is not known. Anyway: the jar files must have a java version lower or equal than the one in the java installation itself. Therefore, an information about these version numbers and the rule during BRouter setup would be certainly helpful.

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp [hunk 1]
```diff
@@ -314,19 +317,109 @@ void CRouterBRouterSetup::readLocalProfiles()
...
+        downloadDir.mkpath(downloadDir.absolutePath());
+    }
+    return downloadDir;
+}
+
+void CRouterBRouterSetup::installLocalBRouter(QStringList& messageList)
+{
+    const QDir targetDir = QDir(localDir);
+    const QDir targetProfileDir = getProfileDir(eModeLocal);
+    const QString srcPath = getDownloadDir().absolutePath();
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp]: fixed - seems I didn't do a lot of c++ recently ;-)

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp [hunk 2]
```diff
@@ -314,19 +317,109 @@ void CRouterBRouterSetup::readLocalProfiles()
...
+    QDirIterator srcIterator(srcPath, {"*.jar","*.brf","lookups.dat"}, QDir::Files, QDirIterator::Subdirectories);
+    QStringList jarFiles;
+    while(srcIterator.hasNext())
+    {
+        QFile srcFile(srcIterator.next());
+        QFileInfo srcFileInfo(srcFile);
+        QString srcFileName = srcFileInfo.fileName();
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp]: fixed

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp [hunk 3]
```diff
@@ -314,19 +317,109 @@ void CRouterBRouterSetup::readLocalProfiles()
...
+        }
+    }
+    if (jarFiles.isEmpty())
+    {
+        throw tr("no brouter-jar found");
+    }
+    else
+    {
+        bool isFirst = true;
+        for (QString jarFile : jarFiles)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp]: fixed

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp [hunk 4]
```diff
@@ -314,19 +317,109 @@ void CRouterBRouterSetup::readLocalProfiles()
...
+                messageList.append(tr("conflicting alternative jar-file %1, go back to previous page to select!").arg(jarFile));
+            }
+        }
+    }
+}
+
+void CRouterBRouterSetup::installLocalBRouterFile(const QFileInfo& srcFileInfo, const QDir& targetDir, QStringList& messageList) const
+{
+    QString srcAbsoluteFilePath = srcFileInfo.absoluteFilePath();
+    QString targetAbsoluteFilePath(targetDir.absoluteFilePath(srcFileInfo.fileName()));
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp]: fixed

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp [hunk 5]
```diff
@@ -314,19 +317,109 @@ void CRouterBRouterSetup::readLocalProfiles()
...
+    if (QFile::exists(targetAbsoluteFilePath))
+    {
+        for (int i=1;;i++)
+        {
+            QString newFilename = targetAbsoluteFilePath+"("+QString::number(i)+")";
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp]: fixed

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.h [hunk 6]
```diff
@@ -164,7 +171,8 @@ private slots:
     const QString defaultBinariesUrl = "https://brouter.de/brouter_bin/";
     const QString defaultSegmentsUrl = "https://brouter.de/brouter/segments4/";
 
-    const QString onlineCacheDir = "BRouter";
+    const QString onlineProfileCacheDir = "BRouterProfiles";
+    const QString downloadCacheDir = "BRouterDownload";
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.h]: fixed

**Comment by ntruchsess** [comment]
Querying Java for the Version seems pretty straightforward (just start java --version and parse the result). Not sure this is 100% portable though. To determine the class-file-version from the jar in a portable way I just need to unzip it and read the signature bytes directly from the class-file. Other tools like 'javap' are most likely not installed on a normal users pc (javap is part of the jdk not the jre)

**Comment by kiozen** [comment]
what a about dummy starting brouter. If it fails it's a java problem.

**Comment by kiozen** [comment]
How to proceed? Do you want to add a Java check? Or should we get the PR into dev first?

**Comment by ntruchsess** [comment]
I've started on the Java-Check and it's 8% done but then new working-week started. I'll finish it on the upcoming weekend

**Comment by ntruchsess** [comment]
I did add determination of java-version and class-file-format version. Both are shown in the dialog. For now I did not add an automatic check whether those versions are actually compatible as I do not know whether the output of 'java --version' is equivalent on all platforms. If it is not it will show 'java version 'unknown' but nevertheless it still might work (we just cannot tell beforehand).
The class-file-version determination should work on all platforms so in case the version of the java executable cannot be determined the user still can see which minimum version is required and may proceed on own risk ;-)

**Comment by ntruchsess** [comment]
had to fine-tune the dialog so it works in both expert and non-expert mode

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp [hunk 7]
```diff
@@ -726,17 +819,99 @@ void CRouterBRouterSetup::loadOnlineProfileFinished(QNetworkReply* reply)
...
+{
+    localBRouterJar = path;
+    const QString& jarFileName = QDir(localDir).absoluteFilePath(localBRouterJar);
+    const QStringList& classFiles = JlCompress::getFileList(jarFileName).filter(QRegularExpression(".*BRouter\\.class"));
+    if (!classFiles.isEmpty())
+    {
+        const QString& tmpFileName = JlCompress::extractFile(jarFileName,classFiles.first(),getDownloadDir().absoluteFilePath("BRouter.class"));
+        if (tmpFileName.endsWith("BRouter.class"))
+        {
+            QFile tmpFile(tmpFileName);
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp:832]: Using `QTemporaryFile` with auto-remove set true would be much safer as you do not have to take care that the file is removed on every exit point.

**Review by kiozen — CHANGES_REQUESTED** [review]
Can we have a line in the wizard that is set visible when the java versions do not match? Something like:
```
    if(setup->javaMajorVersion == NOINT || setup->javaMajorVersion < setup->classMajorVersion)
    {
        const QString& msg = tr(
            "Your Java version %1 seems to be older than the required version %2. "
            "BRouter will probably not work as expected. Please check the logs if "
            "Brouter failed."
            ).arg(setup->javaMajorVersion == NOINT ? tr("unknown") :  QString::number(setup->javaMajorVersion)).arg(setup->classMajorVersion);

        qDebug() << msg;
      // todo: set line visible and set line text to msg. The text of the line might be red to catch the users attention.
    }
```

Maybe there is a better phrasing. 

With my default old Java installation (1.8)  get version unknown as the call to get the version requires "-version" instead of "--version".  I couldn't think of a good solution for that right on the spot. Maybe you have an idea.

File: src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp (hunk 7 above)
Comment (ntruchsess) [reply src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp:832]: I guess QTemporaryFile will not work here as the physical file is created by JlCompress::extractFile before QFile is declared. QTemporaryFile with autoremoval is suitable when the creation of the physical file will be controlled by QTemporaryFile (by calling 'open'). I cannot see how I can make JlCompress return a QTemporaryFile without changing the library (which I do not intent to do)

**Comment by ntruchsess** [comment]
regarding the java-version-determination: I think it makes sense to just test-start java multiple times (once with -version and once with --version) and parse the results until a valid version is found.
Will add this requested line to the dialog, this is no big deal ;-)

**Comment by ntruchsess** [comment]
I did refactor the java-check so further plattform-specific arguments and patterns may be added easily if required:
https://github.com/ntruchsess/qmapshack/blob/5bfad721f26b4ff56acb49971c5ec5b8317a4f82/src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetup.cpp#L889

As suggested if the version is unknown or to low a red warning is now displayed:
https://github.com/ntruchsess/qmapshack/blob/5bfad721f26b4ff56acb49971c5ec5b8317a4f82/src/qmapshack/gis/rte/router/brouter/CRouterBRouterSetupWizard.cpp#L478

From my point of view this is now ready to merge.

**Review by kiozen — APPROVED** [review]

**Pull request merged**